plugins/Kaleidoscope-FocusSerial/src/kaleidoscope/plugin/FocusSerial.cpp

```diff
@@ -77,24 +82,63 @@ EventHandlerResult FocusSerial::afterEachCycle() {
   return EventHandlerResult::OK;
 }
 
+
+bool FocusSerial::inputMatchesCommand(const char *input, const char *to_match) {
+  if (strncmp_P(input, to_match, strlen_P(to_match)) == 0) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool FocusSerial::inputMatchesSubcommand(const char *input, const char *command_to_match, const char *subcommand_to_match) {
+  if (strcmp_P(input + strlen_P(command_to_match), subcommand_to_match) == 0) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
 bool FocusSerial::handleHelp(const char *command,
                              const char *help_message) {
-  if (strcmp_P(command, PSTR("help")) != 0)
+  if (!inputMatchesCommand(command, focus_command_help_))
     return false;
 
   Runtime.serialPort().println((const __FlashStringHelper *)help_message);
   return true;
 }
 
+
+bool FocusSerial::handleHelp(const char *command,
+                             uint8_t help_message_count,
+                             ...) {
+
+  if (!inputMatchesCommand(command, focus_command_help_))
+    return false;
+
+  va_list messages;
+  va_start(messages, help_message_count);
+
+  for (int i = 0; i < help_message_count; ++i) {
+    char *message = va_arg(messages, char *);
+    Runtime.serialPort().print((const __FlashStringHelper *)message);
+  }
+  Runtime.serialPort().println();
+  va_end(messages);
+  return true;
+}
```

Comment:
> Multiple PSTRs for the same string are included in flash multiple times. While that isn't a huge deal for us on GD32, it's still not _good_ and something that can be improved.

Indeed, but if we pull out the *full* command name into variables, we don't necessarily need to include them multiple times. More on that below.


> Thinking more about it, I now wonder if the right solution would instead be to completely redo how we typically handle "help" and instead of making the plugin respond to help, we teach plugins how to offer up a list of all the commands they understand as flash strings and let Focus synthesize help output from those lists.

I think that my proposed API [here](https://github.com/keyboardio/Kaleidoscope/blob/f/handleHelp-varargs/plugins/Kaleidoscope-FocusSerial/src/kaleidoscope/plugin/FocusSerial.h#L52-L59) does something very close to that.

That transforms the `if (handleHelp(<list of commands>)) return OK` pattern into `if (command == "help") return printHelp(<list of commands>)`.

The problem with lifting out the command list from `onFocusEvent` is that we want to have access to the variables in that function too, to avoid repetition. However, now that I'm trying to explain why we can't... I have an idea!

What if we'd put the commands into an array within the plugin class, and we'd have an enum or the like also in there, to help address the command PSTRs from `onFocusEvent`? Then `Focus` could just iterate over the array until it finds a sentinel, and the individual `onFocusEvent`s could pick out the command they need. This needs a little bit of care to maintain the mapping, but perhaps we could write a preprocessor macro for that?

In this case, `onFocusEvent` would not deal with help *at all*, saving us a lot of bytes already. It would just do:

```c++
if (::Focus.inputMatchesCommand(command, focus_commands[<index>])) { 
  /* do stuff */
}
```

Doing this would also be a big leap towards my desire to remove the pattern where we handle the prefix and subcommands separately: we'd just use the full strings, which would be stored in one place, one copy of them.

> Those lists could then also be used internally by the plugins instead of repeating textual strings in command processing.

We don't necessarily need to move the strings out of `onFocusEvent()` for that, see [here](https://github.com/keyboardio/Kaleidoscope/blob/f/handleHelp-varargs/plugins/Kaleidoscope-FocusSerial/src/kaleidoscope/plugin/FocusSerial.h#L52-L59):

```c++
EventHandlerResult FocusSerial::onFocusEvent(const char *command) {
  const char *cmd_help    = PSTR("help");
  const char *cmd_reset   = PSTR("device.reset");
  const char *cmd_plugins = PSTR("plugins");

  if (isHelp(command))
    return printHelp(cmd_help, cmd_reset, cmd_plugins);

  if (strcmp_P(command, cmd_reset) == 0) {
    Runtime.device().rebootBootloader();
    return EventHandlerResult::EVENT_CONSUMED;
  }
  if (strcmp_P(command, cmd_plugins) == 0) {
    kaleidoscope::Hooks::onNameQuery();
    return EventHandlerResult::EVENT_CONSUMED;
  }

  return EventHandlerResult::OK;
}
```

(Imagine this uses the `::Focus.inputMatchesCommand()` helper instead of `strcmp_P`, and a better name for `isHelp()`)

Always uses the full strings, does not check the prefix first, and then only suffixes, it always uses the same full string. So there exists only one copy. We only have one copy of `help` too, because that's checked in a helper function.

Reply:
Sounds plausible.